### DISCUSSION TITLE
Normalising data received in `CommandSettings`

### DISCUSSION THREAD

**Discussion by lonix1** [spectreconsole/spectre.console#933] [Q&A] [answered]
I have a `CommandSettings` subclass with data that should be pre-processed / normalised:

````cs
public sealed class MySettings : CommandSettings
{
  [CommandOption("-p|--path")]
  public string? Path { get; init; }
}
````

Let's say I receive `~/foo` - then I must expand it to `/home/mike/foo`.

Does `CommandSettings` have some hook I could use to perform a "cleaning" / "normalising" / "preprocessing" step?

**Comment by patriksvensson** [comment, marked as answer]
You can either do this via a custom type converter, or via the `Validate` method on either the settings class or the command.

https://spectreconsole.net/cli/settings
https://spectreconsole.net/cli/commands

**Reply by lonix1** [reply]
Thanks Patrik. I found another way for anyone else who has this need. I added the cleaning in the ctor:

````cs
public class Settings : CommandSettings
{

  public Settings(string path)
  {
    Path = Clean(path);
  }

  [CommandOption("-p|--path")]
  public string? Path { get; init; }

}
````

(Which relies on a very useful binding feature - thanks for designing it that way.)

**Reply by patriksvensson** [reply]
Oh, yeah, I forgot about that option 😅. That’s probably the most intuitive one.
